### PR TITLE
73387 upload attachment to tag

### PR DESCRIPTION
**Pull request by janid1967**

### CODE REVIEW

File: src/Equinor.Procosys.Preservation.Command/TagAttachmentCommands/Upload/UploadTagAttachmentCommandHandler.cs [hunk 1]
```diff
@@ -14,45 +15,58 @@ public class UploadTagAttachmentCommandHandler : IRequestHandler<UploadTagAttach
...
             _plantProvider = plantProvider;
+            _blobStorage = blobStorage;
+            _attachmentOptions = attachmentOptions;
         }
 
         public async Task<Result<int>> Handle(UploadTagAttachmentCommand request, CancellationToken cancellationToken)
         {
             var tag = await _projectRepository.GetTagByTagIdAsync(request.TagId);
 
-            var attachment = tag.GetAttachmentByFileName(request.FileName);
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/TagAttachmentCommands/Upload/UploadTagAttachmentCommandHandler.cs:29]: throw exception if .File not given

File: src/Equinor.Procosys.Preservation.Command/TagAttachmentCommands/Upload/UploadTagAttachmentCommandHandler.cs [hunk 2]
```diff
@@ -14,45 +15,58 @@ public class UploadTagAttachmentCommandHandler : IRequestHandler<UploadTagAttach
...
+            var path = GetPath(nameof(Tag), attachment);
+
+            await _blobStorage.UploadAsync(path, request.File.OpenReadStream(), request.OverwriteIfExists, cancellationToken);
+
             await _unitOfWork.SaveChangesAsync(cancellationToken);
 
             return new SuccessResult<int>(attachment.Id);
         }
+
+        private string GetPath(string folderName, Attachment attachment)
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/TagAttachmentCommands/Upload/UploadTagAttachmentCommandHandler.cs:69]: Must be available for Queries too. Need it in download

File: src/Equinor.Procosys.Preservation.Command/TagAttachmentCommands/Upload/UploadTagAttachmentCommandValidator.cs [hunk 3]
```diff
@@ -14,15 +14,16 @@ public class UploadTagAttachmentCommandValidator : AbstractValidator<UploadTagAt
...
             RuleFor(command => command)
                 .MustAsync((command, token) => NotBeAClosedProjectForTagAsync(command.TagId, token))
                 .WithMessage(command => $"Project for tag is closed! Tag={command.TagId}")
                 .MustAsync((command, token) => BeAnExistingTagAsync(command.TagId, token))
                 .WithMessage(command => $"Tag doesn't exist! Tag={command.TagId}")
                 .MustAsync((command, token) => NotBeAVoidedTagAsync(command.TagId, token))
                 .WithMessage(command => $"Tag is voided! Tag={command.TagId}")
-                .MustAsync((command, token) => NotHaveAttachmentWithFilenameAsync(command.TagId, command.FileName, token))
-                .WithMessage(command => $"Tag already have an attachment with filename {command.FileName}! Please rename file or choose to overwrite")
+                .MustAsync((command, token) => NotHaveAttachmentWithFilenameAsync(command.TagId, command.File.FileName, token))
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Command/TagAttachmentCommands/Upload/UploadTagAttachmentCommandValidator.cs:25]: .when .File != null

File: src/Equinor.Procosys.Preservation.Domain/Attachment.cs [hunk 4]
```diff
@@ -14,22 +14,29 @@ protected Attachment()
...
         public Guid BlobStorageId { get; private set; }
-        public string Title { get; set; }
+        public string Title { get; private set; }
         public string FileName { get; private set; }
         public DateTime CreatedAtUtc { get; private set; }
         public int CreatedById { get; private set; }
         public DateTime? ModifiedAtUtc { get; private set; }
         public int? ModifiedById { get; private set; }
 
+        public void SetTitle(string title, string fileName) => Title = title ?? fileName;
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.Domain/Attachment.cs:38]: add tests

File: src/Equinor.Procosys.Preservation.WebApi/Equinor.Procosys.Preservation.WebApi.csproj [hunk 5]
```diff
@@ -8,6 +8,10 @@
     <ApplicationInsightsAnnotationResourceId>/subscriptions/343069e3-ceaa-4256-bdc8-49fa3646c1c4/resourcegroups/pcs-preservation-dev-rg/providers/microsoft.insights/components/pcs-pres-dev-ai</ApplicationInsightsAnnotationResourceId>
   </PropertyGroup>
 
+  <ItemGroup>
+    <Compile Remove="Swagger\AddUploader.cs" />
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.WebApi/Equinor.Procosys.Preservation.WebApi.csproj:12]: remove

File: src/Equinor.Procosys.Preservation.WebApi/Swagger/AddUploader.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Http;
+using Microsoft.OpenApi.Models;
+using Swashbuckle.AspNetCore.SwaggerGen;
+using Swashbuckle.AspNetCore.Swagger;  
+
+namespace Equinor.Procosys.Preservation.WebApi.Swagger
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.WebApi/Swagger/AddUploader.cs:9]: remove class

File: src/Equinor.Procosys.Preservation.WebApi/Swagger/ApiDescriptionExtension.cs [hunk 7]
```diff
@@ -14,5 +14,8 @@ public static AuthorizeAttribute[] GetAuthorizeAttributes(this ApiDescription ap
 
             return hasAuthAttribute as AuthorizeAttribute[] ?? hasAuthAttribute.ToArray();
         }
+
+        //public static UploadAttribute GetUploadAttribute(this ApiDescription apiDescription)
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.WebApi/Swagger/ApiDescriptionExtension.cs:18]: remove

File: src/Equinor.Procosys.Preservation.WebApi/appsettings.json [hunk 8]
```diff
@@ -30,6 +30,10 @@
   "TagOptions": {
     "TagIsNewHours": 48
   },
+  "AttachmentOptions": {
+    "MaxSizeKb": 4000,
+    "BlobContainer": "Preservation"
```
Comment (janid1967) [inline src/Equinor.Procosys.Preservation.WebApi/appsettings.json:35]: lowercase

File: src/Equinor.Procosys.Preservation.Domain/Attachment.cs (hunk 4 above)
Comment (Cnordbo) [inline src/Equinor.Procosys.Preservation.Domain/Attachment.cs:38]: This should probably just accept a single string, and have a setter that validates that the new title is not empty. 
The logic for setting fileName as the default title should probably be defined in the constructor.
```js
Title = title || fileName

**Pull request closed without merging**